src/hotspot/share/opto/cfgnode.cpp

```diff
@@ -1779,6 +1779,70 @@ static Node* is_absolute( PhaseGVN *phase, PhiNode *phi_root, int true_path) {
   return x;
 }
 
+static Node* is_minmax(PhaseGVN* phase, PhiNode* phi_root, int true_path) {
+  assert(true_path != 0, "only diamond shape graph expected");
+
+  Node* region = phi_root->in(0);
+  IfNode* iff = region->in(1)->in(0)->as_If();
+  BoolNode* bol = iff->in(1)->as_Bool();
+  Node* cmp = bol->in(1);
+
+  // Don't transform if the branch is highly predictable
+  constexpr float infrequent_prob = PROB_UNLIKELY_MAG(2);
+  if (iff->_prob < infrequent_prob || iff->_prob > (1.0f - infrequent_prob)) {
+    return nullptr;
+  }
+
+  int false_path = 3 - true_path;
+  bool is_min = false;
+  int opcode = cmp->Opcode();
+  bool is_long = opcode == Op_CmpL;
+
+  // Ensure comparison is an integral type
+  if (opcode == Op_CmpI || opcode == Op_CmpL) {
+    int test = bol->_test._test;
+    // Only accept canonicalized le and lt comparisons
+    if (test != BoolTest::le && test != BoolTest::lt) {
+      return nullptr;
+    }
+  } else {
+    return nullptr;
+  }
+
+  Node* cmp_true = cmp->in(true_path);
+  Node* cmp_false = cmp->in(false_path);
+
+  // Identify which path would result in a minimum being created for "a < b ? a : b" and "a < b ? b : a"
+  // If neither structure is found, exit
+  int min_path;
+  if ((cmp_true == phi_root->in(true_path)) && (cmp_false == phi_root->in(false_path))) {
+    min_path = 1;
+  } else if (cmp_true == phi_root->in(false_path) && cmp_false == phi_root->in(true_path)) {
+    min_path = 2;
+  } else {
+    return nullptr;
+  }
```

Comment:
Maybe also have variables `phi_true/false`, would make the code easier to read.

Reply:
Hmm. Maybe you can refactor this, or explain somehow why exactly the conditions lead to `min_path = 1/2`

Reply:
I've added new variables for `phi_true` and `phi_false`, and also added a comment explaining how those values are selected.